enhancements/api-review/azure_user_defined_tags.md

```diff
@@ -0,0 +1,322 @@
+`azurerm_resource_group, azurerm_image, azurerm_lb, azurerm_network_security_group, azurerm_storage_account, azurerm_user_assigned_identity, azurerm_virtual_network, azurerm_linux_virtual_machine, azurerm_network_interface, azurerm_dns_cname_record`
+
+API update example:
+A local variable should be defined, which merges the default tag and the user
+defined Azure tags, which should be referred in the Azure resource APIs.
+``` terraform
+locals {
+  tags = merge(
+    {
+      "kubernetes.io_cluster.${var.cluster_id}" = "owned"
+    },
+    var.azure_extra_tags,
+  )
+}
+
+resource "azurerm_resource_group" "main" {
+  tags     = local.tags
+}
+```
+
+#### Caveats
+1. User updates or removes Azure resource tags added by Openshift, using an external interface
+   will not be reconciled by the resource owning operator.
+2. Updating tags of individual resources is not supported and any tag present in 
+   `.status.platformStatus.azure.resourceTags` of `infrastructure.config.openshift.io/v1` resource 
+   will result in addding tags to all Openshift managed Azure resources. 
+
+### Risks and Mitigations
+
+### Drawbacks
+- User-defined tags cannot be updated on an Azure resource which is not managed by an
+  operator. In this proposal, the changes proposed and developed will be part of
+  openshift-* namespace. External operators are not in scope.
+  User-defined tags can be updated on the following Azure resources.
+    1. Virtual machine resources created for master and worker nodes.
+    2. Storage account
+    3. NetworkEdge
+    4. Internal Registry
+    5. CCO
+
+- OpenShift is bound to have the common limitation for all Azure resources created
+  by it and constraints other resources with the least matching limit as below
+    1. Tag names cannot have `microsoft`, `azure`, `windows` prefixes which are
+       reserved for Azure use.
+    2. An Azure storage account has a limit of 128 characters for the tag name.
+    3. An Azure DNS zone or Traffic or Front Door resource tag name cannot have spaces,
+       special/unicode characters or start with a number.
+    4. An Azure Automation or Content Delivery Network or DNS resource can have a
+       maximum of 15 tags.
+
+- Administrator will have to manually perform below tags pertaining actions
+    1. removing the undesired tags from the required resources.
+    2. update tag values of the required resources.
+    3. update tags of the resources which are not managed by an operator.
+    4. update tags of the resources for which update logic is not supported by an operator.
+
+## Design Details
+
+### Open Questions
+
+### Test Plan
+
+### Graduation Criteria
+
+#### Dev Preview -> Tech Preview
+
+#### Tech Preview -> GA
+
+#### Removing a deprecated feature
+
+### Upgrade / Downgrade Strategy
+
+On upgrade:
+- Cluster operators that update the tags of Azure resources created for cluster 
+  should refer the new fields and take action. 
```

Comment:
the implication here is that on upgrade we'd expect clusteroperators to read status.resourceTags and then update the tags on existing azure resources?

is that a requirement?  For AWS we only committed to setting the tags on AWS resources at creation time, right?  Updating existing cloud resources w/ new/updated tags was "unspecified" behavior that very much depends on what the underlying resource supports, and what the component managing the resource decided to do.

Reply:
Yeah, even for Azure, tags will be added to the resources during creation time. But if the operators managing the Azure resources are upgraded, if any tags are present in resourceTags those will be added to resources by such operators which reconcile resource properties.

Reply:
> . But if the operators managing the Azure resources are upgraded, if any tags are present in resourceTags those will be added to resources by such operators which reconcile resource properties.

right, so it's not a requirement, it's unspecified behavior.  Some operators will reconcile the tags, others may not.  I guess that's what you're saying with "Cluster operators that update tags", but i think it's worth calling out the unspecified nature of this behavior more explicitly.

Reply:
comment incorporated